### PR TITLE
[cudamapper] Grouping reads into indices moved in cudamapper

### PR DESCRIPTION
**Pull request by mimaric**
Grouping reads into indices was done in `FastaParser::get_read_chunks()`. The function is somewhat related to `FastaParser`, but it is cudamapper-specific and one could argue that is has more to with `Index` than `FastaParser`, so now it is a free function called `group_reads_into_indices()` in cudamapper.

Extracted `IndexDescriptor` into a separate file (together with `group_reads_into_indices()`). `IndexDescriptor` is only vaguely related to `IndexCache`.

Improved types in `FastaParser`.

Created tests for `group_reads_into_indices()`

Part of issue #318 

@tijyojwad FYI

### CODE REVIEW

File: common/io/include/claragenomics/io/fasta_parser.hpp [hunk 1]
```diff
@@ -62,7 +54,9 @@ class FastaParser
 /// \param shuffle Enables shuffling reads
 ///
 /// \return A unique pointer to a constructed parser object.
-std::unique_ptr<FastaParser> create_kseq_fasta_parser(const std::string& fasta_file, int min_sequence_length = 0, bool shuffle = true);
+std::unique_ptr<FastaParser> create_kseq_fasta_parser(const std::string& fasta_file,
+                                                      const number_of_basepairs_t min_sequence_length = 0,
```
Comment (ahehn-nv) [inline common/io/include/claragenomics/io/fasta_parser.hpp:58]: I wouldn't put `const` in the function declaration for call-by-value parameters. It is call-by-value, so the values will be copied - that's the semantic of this function. The constness is an implementation detail, so I would add it only into the implementation of the function. The compiler should fine with this difference.

File: common/io/src/kseqpp_fasta_parser.hpp [hunk 2]
```diff
@@ -22,20 +22,28 @@ namespace io
...
-
-    FastaSequence get_sequence_by_id(int32_t i) const override;
-
-    std::vector<std::pair<int, int>> get_read_chunks(int max_chunk_size) const override;
+    /// \brief Constructor
+    /// \param fasta_file Path to FASTA(.gz) file. If .gz, it must be zipped with bgzip.
+    /// \param min_sequence_length Minimum length a sequence needs to be to be parsed. Shorter sequences are ignored.
+    /// \param shuffle Enables shuffling reads
+    FastaParserKseqpp(const std::string& fasta_file,
+                      const number_of_basepairs_t min_sequence_length,
```
Comment (ahehn-nv) [inline common/io/src/kseqpp_fasta_parser.hpp:30]: `const` call-by-value in declaration (see above).

File: common/io/src/kseqpp_fasta_parser.hpp [hunk 3]
```diff
@@ -22,20 +22,28 @@ namespace io
...
+                      const bool shuffle);
+
+    /// \brief Return number of sequences in FASTA file
+    /// \return Sequence count in file
+    number_of_reads_t get_num_seqences() const override;
+
+    /// \brief Fetch an entry from the FASTA file by index position in file.
+    /// \param sequence_id Position of sequence in file. If sequence_id is invalid an error is thrown.
+    /// \return A FastaSequence object describing the entry.
+    FastaSequence get_sequence_by_id(const read_id_t sequence_id) const override;
```
Comment (ahehn-nv) [inline common/io/src/kseqpp_fasta_parser.hpp:40]: `const` call-by-value in declaration (see above).

File: cudamapper/src/index_descriptor.hpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+};
+
+/// \brief returns a list of IndexDescriptors in which the sum of basepairs of all reads in one IndexDescriptor is at most max_basepairs_per_index
+/// If a single read exceeds max_chunk_size it will be placed in its own IndexDescriptor.
+///
+/// \param parser parser to get the reads from
+/// \param max_basepairs_per_index the maximum number of basepairs in an IndexDescriptor
+/// \return list of IndexDescriptors
+std::vector<IndexDescriptor> group_reads_into_indices(const io::FastaParser& parser,
+                                                      const number_of_basepairs_t max_basepairs_per_index = 1000000);
```
Comment (ahehn-nv) [inline cudamapper/src/index_descriptor.hpp:82]: `const` call-by-value in declaration (see above).

File: cudamapper/tests/Test_CudamapperIndexDescriptor.cpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+{
+namespace cudamapper
+{
+
+// *** test IndexDescriptor and IndexDescriptorHash ***
+
+TEST(TestCudamapperIndexDescriptor, test_index_descriptor_getters)
+{
+    const read_id_t first_read      = 15;
+    const read_id_t number_of_reads = 156;
```
Comment (ahehn-nv) [inline cudamapper/tests/Test_CudamapperIndexDescriptor.cpp:27]: I guess that should be a `number_of_reads_t` now. :)

File: cudamapper/tests/Test_CudamapperIndexDescriptor.cpp [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    ASSERT_NE(index_descriptor_15_156_1, index_descriptor_16_156);
+    ASSERT_NE(index_descriptor_15_156_1, index_descriptor_15_157);
+    ASSERT_NE(index_descriptor_15_156_1, index_descriptor_16_157);
+}
+
+TEST(TestCudamapperIndexDescriptor, test_index_descriptor_hash)
+{
+    static_assert(sizeof(size_t) == 8, "only 64-bit values supported, adjust element_mask and shift_bits");
+    const read_id_t first_read      = 0x24;
+    const read_id_t number_of_reads = 0xCF;
```
Comment (ahehn-nv) [inline cudamapper/tests/Test_CudamapperIndexDescriptor.cpp:52]: `number_of_reads_t`?

File: cudamapper/src/index_descriptor.cpp [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+{
+    generate_hash();
+}
+
+read_id_t IndexDescriptor::first_read() const
+{
+    return first_read_;
+}
+
+read_id_t IndexDescriptor::number_of_reads() const
```
Comment (ahehn-nv) [inline cudamapper/src/index_descriptor.cpp:33]: `number_of_reads_t`?

File: cudamapper/src/index_descriptor.cpp [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+{
+    static_assert(sizeof(std::size_t) == 8, "only 64-bit values supported, adjust element_mask and shift_bits");
+
+    // populate lower half of hash with one value, upper half with the other
+    constexpr std::size_t element_mask = 0xFFFFFFFF; // for 64 bits
+    constexpr std::uint32_t shift_bits = 32;         // for 64 bits
+
+    hash_ = 0;
+    hash_ |= first_read_ & element_mask;
+    hash_ |= static_cast<std::size_t>(number_of_reads_ & element_mask) << shift_bits;
```
Comment (ahehn-nv) [inline cudamapper/src/index_descriptor.cpp:53]: Not part of this PR, but I guess the `static_cast` should only cast `number_of_reads_`?

File: cudamapper/src/main.cu [hunk 9]
```diff
@@ -450,25 +451,26 @@ int main(int argc, char* argv[])
 
     struct QueryTargetsRange
     {
-        std::pair<std::int32_t, int32_t> query_range;
-        std::vector<std::pair<std::int32_t, int32_t>> target_ranges;
+        claragenomics::cudamapper::IndexDescriptor query_range;
```
Comment (ahehn-nv) [inline cudamapper/src/main.cu:454]: Optional: I think I would put a `using claragenomics::cudamapper::IndexDescriptor;` at the top of `main()`, to make the following code a tiny bit more readable.

File: common/io/include/claragenomics/io/fasta_parser.hpp (hunk 1 above)
Comment (mimaric) [reply common/io/include/claragenomics/io/fasta_parser.hpp:58]: Cool, I didn't know that. Thanks!
Done

File: common/io/src/kseqpp_fasta_parser.hpp (hunk 2 above)
Comment (mimaric) [reply common/io/src/kseqpp_fasta_parser.hpp:30]: done

File: common/io/src/kseqpp_fasta_parser.hpp (hunk 3 above)
Comment (mimaric) [reply common/io/src/kseqpp_fasta_parser.hpp:40]: done

File: cudamapper/src/index_descriptor.hpp (hunk 4 above)
Comment (mimaric) [reply cudamapper/src/index_descriptor.hpp:82]: done

File: cudamapper/tests/Test_CudamapperIndexDescriptor.cpp (hunk 5 above)
Comment (mimaric) [reply cudamapper/tests/Test_CudamapperIndexDescriptor.cpp:27]: Yes.
There are also many other places which will benefit from this, leaving those as they are for now.
Done

File: cudamapper/tests/Test_CudamapperIndexDescriptor.cpp (hunk 6 above)
Comment (mimaric) [reply cudamapper/tests/Test_CudamapperIndexDescriptor.cpp:52]: done

File: cudamapper/src/index_descriptor.cpp (hunk 7 above)
Comment (mimaric) [reply cudamapper/src/index_descriptor.cpp:33]: done

File: cudamapper/src/index_descriptor.cpp (hunk 8 above)
Comment (mimaric) [reply cudamapper/src/index_descriptor.cpp:53]: yes, done

File: cudamapper/src/main.cu (hunk 9 above)
Comment (mimaric) [reply cudamapper/src/main.cu:454]: My plan is to have the whole `main.cu` in namespace `claragenomcis::cudamapper`. I don't see a reason how it to not be in it. I'm planning to do this in a separate PR

File: common/io/include/claragenomics/io/fasta_parser.hpp (hunk 1 above)
Comment (ahehn-nv) [reply common/io/include/claragenomics/io/fasta_parser.hpp]: sure, you're welcome. :)

File: cudamapper/src/main.cu (hunk 9 above)
Comment (ahehn-nv) [reply cudamapper/src/main.cu:454]: ok.
